### PR TITLE
Feature/game socket

### PR DESCRIPTION
**Pull request by lhossackucalgary**
Add nuxt-socket-io module w/ socket authentication server & client. 
Socket init in navbar, can commit to vuex store.

I'll add game events/ mutations tonight, likely will move io to separate config

### CODE REVIEW

**Comment by lhossackucalgary** [comment]
Scope of this PR is starting to get a little out of hand; hopefully it is still legible. 

Overview: 
- Add socket authentication & vuex sockets
- Organize/ refactor backend into not 1 file
- Configure sockets & api authentication
- Fix double-connection socket issues

I was hoping to get game api & socket events + plug into game engine implemented, but it doesn't look like I'll have time for this until Sunday (Nov. 29)

File: batan-client/components/navbar.vue [hunk 1]
```diff
@@ -39,36 +39,77 @@ export default Vue.extend({
...
+        this.$root.gcSock.disconnect()
+      } catch (err) {
+        // ignore.. socket already closed
+      }
+
+      // Log out - clearing
+      this.$auth.logout();
+
+      // TODO: Clear all local user data (vuex gameboard, chat, localStorage, cookies)..
+      // maybe just refresh page after logout?
```
Comment (rk4514) [inline batan-client/components/navbar.vue]: Not sure if refreshing the page will clear the cookies, but we can deal with that when we get to it.

File: batan-client/store/games.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+var game_mock = {
+  "game 1": {"id": "longUID", "board": {}},
+  "game 2": {"id": "longUID2", "board": {}}
+}
```
Comment (rk4514) [inline batan-client/store/games.js:4]: Is this for the client having two games going at once? Do we have to limit to two?

**Review by rk4514 — APPROVED** [review]
Looks good to me, however I'm still not super familiar with VUE so if someone else wants to take a look as well that would be appreciated. Just a couple questions from me but overall makes sense.

File: batan-client/store/games.js (hunk 2 above)
Comment (lhossackucalgary) [reply batan-client/store/games.js:4]: This is a mock so I could check the navbar; the idea is to have events populate the whole data structure with limits imposed explicitly, if at all

File: batan-client/components/navbar.vue (hunk 1 above)
Comment (lhossackucalgary) [reply batan-client/components/navbar.vue]: Sorry for the confusing comment - the cookies & local storage will not be cleared on refresh, but the vuex store is reset. Session cookies should be cleared by $auth.logout()

**Comment by jackadrianglass** [comment]
It looks good to me. The only thing that I can think of is maybe combing through and removing any commented out code that you don't expect to need later on. But that's not a big deal

File: batan-client/components/navbar.vue (hunk 1 above)
Comment (lhossackucalgary) [reply batan-client/components/navbar.vue]:
```suggestion
```

File: batan-client/pages/test_sockets.vue [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      })
+      // Couldn't get this to work, nuxt-socket-io docs may be outdated
+      // await this.$store.dispatch(
+      //   '$nuxtSocket/emit',
+      //   {
+      //     label: 'gcSock',
+      //     evt: 'hello',
+      //     msg: { hi: 'world' }
+      //   }
+      // )
```
Comment (lhossackucalgary) [inline batan-client/pages/test_sockets.vue]:
```suggestion
```

File: batan-server/app/api-http/routes/test.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Router, Request, Response } from 'express';
+const route = Router();
+
+export default (app: Router) => {
+  app.use('/test', route);
+
+  route.get('/hi', function (req: Request, res: Response) {
+    // Connections to the API are authenticated using Authorization header &
+    // Bearer jwt strategy. Access user's UID by req.user.sub
+    // console.log(`From https: ${req.user.sub}`);
```
Comment (lhossackucalgary) [inline batan-server/app/api-http/routes/test.ts]:
```suggestion
```

File: batan-server/app/api-socket/events/connection.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import socketState from '../../state/socket';
+
+export default (io, socket) => {
+  socketState.addSocket(socket.decoded_token.sub, socket.id);
+  // console.log(socketState.online);
```
Comment (lhossackucalgary) [inline batan-server/app/api-socket/events/connection.ts]:
```suggestion
```

File: batan-server/app/api-socket/events/connection.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+  socket.on('disconnect', (reason) => {
+    socketState.rmSocket(socket.decoded_token.sub, socket.id);
+    // console.log(socketState.online);
```
Comment (lhossackucalgary) [inline batan-server/app/api-socket/events/connection.ts]:
```suggestion
```

File: batan-server/app/api-socket/events/test.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+export default (io, socket) => {
+  // console.log(`From socket: ${socket.decoded_token.sub}`);
```
Comment (lhossackucalgary) [inline batan-server/app/api-socket/events/test.ts]:
```suggestion
```

File: batan-server/app/state/game.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+games: game_id -> {
+  players: [
+    (uid1, color1),
+    (uid2, color2),
+    ..
+  ],
+  game_owner: uid,
+  game_obj: obj
+}
+*/
```
Comment (lhossackucalgary) [inline batan-server/app/state/game.ts]:
```suggestion

```

**Pull request merged**